File: cpp/prob001.cpp
#include <iostream>
#include <chrono>
#include "utils.h"

#define PROB_NUM    2

int64_t compute()
{
    int sum = 0;
    for(int i = 0; i < 1000; i++)
    {
        if(i%3 == 0 or i%5 == 0)
            sum += i;
    }
    return sum;
}


int test_answer(std::string prob_answer) {
    std::string answer;
    if(!get_answer(PROB_NUM, &answer)) {
        std::cout << "file not found" << std::endl;
        return 1;
    }

    assert(answer.compare(prob_answer) == 0);
    return 0;
}

int main(int argc, char const *argv[])
{
    auto t1 = std::chrono::high_resolution_clock::now();
    int64_t result = compute();
    auto t2 = std::chrono::high_resolution_clock::now();

    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count();
    std::cout << result << "\tin " << duration << " us\n";

    // do testing
    test_answer(std::to_string(result));
    return 0;
}

File: rust/src/bin/prob001.rs
extern crate euler;

fn compute() -> i32{
    let bound = 1000;
    (1..bound).filter(|x| x % 3 == 0 || x % 5 == 0).sum()
}

fn main() {
    let now = std::time::Instant::now();
    let result = compute();
    println!("{:?}\tin {}us", result, now.elapsed().as_micros());
}


#[cfg(test)]
mod tests {
    use super::*;
    const PROB_NUM: usize = 1;
    
    #[test]
    fn test_prob001() {
        let result = compute().to_string();
        euler::test_answer(PROB_NUM, result);
    }
}